### PR TITLE
gh-102211: Document `re.{Pattern,Match}`’s existence

### PR DESCRIPTION
**Pull request by flying-sheep**
`typing` documents `re.Pattern`’s and `re.Match`’s existence so `re` should too.

I’m not sure what to add to the body of the classes, so for now I just added them as markers that indeed, `re` exports two classes by that name.

* Issue: gh-102211

Rendered:

- [`re.Match`](https://deploy-preview-102212--python-cpython-preview.netlify.app/library/re.html#match-objects)
- [`re.Pattern`](https://deploy-preview-102212--python-cpython-preview.netlify.app/library/re.html#regular-expression-objects)

### CODE REVIEW

File: Doc/library/re.rst [hunk 1]
```diff
@@ -1118,6 +1118,8 @@ Regular Expression Objects
 Compiled regular expression objects support the following methods and
 attributes:
 
+.. class: Pattern
```
Comment (JelleZijlstra) [inline Doc/library/re.rst]:
```suggestion
.. class:: Pattern
```
Right now I believe you're adding a comment

File: Doc/library/re.rst [hunk 2]
```diff
@@ -1256,6 +1258,7 @@ when there is no match, you can test whether there was a match with a simple
 
 Match objects support the following methods and attributes:
 
+.. class: Match
```
Comment (flying-sheep) [inline Doc/library/re.rst]:
```suggestion
.. class:: Match
```

**Comment by hugovk** [comment]
Preview: 

image

https://deploy-preview-102212--python-cpython-preview.netlify.app/library/re.html#regular-expression-objects

image

https://deploy-preview-102212--python-cpython-preview.netlify.app/library/re.html#match-objects

---

Should "the following methods and attributes" be changed to "the following classes, methods and attributes"?

> I’m not sure what to add to the body of the classes, so for now I just added them as markers that indeed, `re` exports two classes by that name.

Would be nice to add something rather than leaving blank. Why do these classes exist? What use do they have for end users? "Useful for type hints"?

**Comment by flying-sheep** [comment]
I mean, they’re classes. They can be used for instance checks like all other classes. Weirdly they can’t be instantiated (why not, shouldn’t `re.Pattern(...)` behave the same way as `re.compile(...)`?)

We could also move things up:

> ## Match objects
> 
> *class* re.**Pattern**
> > Compiled regular expression object. See :class:`typing.Pattern` for use in type hints.
> 
> Pattern.**search**(...)
> > ...

**Comment by flying-sheep** [comment]
what would you prefer?

**Comment by hugovk** [comment]
If we're going to add the headers, let's add at least some sort of description, however short.

**Comment by flying-sheep** [comment]
this should be ready now

**Review by hugovk — APPROVED** [review]
Thanks!

File: Doc/library/re.rst [hunk 3]
```diff
@@ -1268,8 +1267,10 @@ when there is no match, you can test whether there was a match with a simple
    if match:
        process(match)
 
-Match objects support the following methods and attributes:
+.. class:: Match
 
+   Match object returned by successful ``match``\ es and ``search``\ es.
+   Supports the following methods and attributes:
```
Comment (AA-Turner) [inline Doc/library/re.rst]: Consistency with the header for ``Pattern``
```suggestion
   It supports the following methods and attributes:
```

File: Doc/library/re.rst [hunk 4]
```diff
@@ -1129,16 +1127,17 @@ Exceptions
 Regular Expression Objects
 --------------------------
 
-Compiled regular expression objects support the following methods and
-attributes:
+.. class:: Pattern
+
+   Compiled regular expression object returned by :func:`re.compile`.
+   It supports the following methods and attributes:
```
Comment (AA-Turner) [inline Doc/library/re.rst]: Perhaps we could add a note about GenericAlias support, e.g.:

```suggestion

   .. versionadded:: 3.9
      :py:class:`re.Pattern` supports ``[]`` to indicate a Unicode (str) or bytes pattern.
      See :ref:`types-genericalias`.
```

**Review by AA-Turner — APPROVED** [review]
Looks good, two comments:

A

File: Doc/library/re.rst [hunk 5]
```diff
@@ -1129,16 +1127,20 @@ Exceptions
 Regular Expression Objects
 --------------------------
 
-Compiled regular expression objects support the following methods and
-attributes:
+.. class:: Pattern
+
+   Compiled regular expression object returned by :func:`re.compile`.
+
+   .. versionadded:: 3.9
```
Comment (AA-Turner) [inline Doc/library/re.rst]:
```suggestion
   .. versionchanged:: 3.9
```

File: Doc/library/re.rst [hunk 6]
```diff
@@ -1268,8 +1270,10 @@ when there is no match, you can test whether there was a match with a simple
    if match:
        process(match)
 
-Match objects support the following methods and attributes:
+.. class:: Match
 
+   Match object returned by successful ``match``\ es and ``search``\ es.
+   It supports the following methods and attributes:
```
Comment (AA-Turner) [inline Doc/library/re.rst]:
```suggestion

   .. versionchanged:: 3.9
      :py:class:`re.Match` supports ``[]`` to indicate a Unicode (str) or bytes match.
      See :ref:`types-genericalias`.
```

**Review by AA-Turner — COMMENTED** [review]
Sorry, I should have made the same suggestion on GenericAlias for both Pattern and Match: (also switched to versionchanged)

**Comment by flying-sheep** [comment]
happy with those changes!

**Review by AA-Turner — APPROVED** [review]

**Comment by flying-sheep** [comment]
Is this waiting for anything?

**Comment by hugovk** [comment]
I think we're good here, thank you for your patience and the contribution!

**Pull request merged**